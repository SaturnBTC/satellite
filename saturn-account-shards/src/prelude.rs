//! Public aliases and helper macro for `ShardSet` ergonomics.
//!
//! Import this as `use saturn_account_shards::prelude::*` to gain the
//! user-friendly fa√ßade without pulling in the full type machinery.

use crate::shard_set::{Selected, ShardSet, Unselected};
use anchor_lang::prelude::AccountLoader;
use anchor_lang::Owner;
use anchor_lang::ZeroCopy;

/// Convenience alias for an **unselected** `ShardSet` in which the typestate
/// parameter is hidden.
pub type Shards<'slice, 'info, S, const MAX_SELECTED_SHARDS: usize> =
    ShardSet<'slice, 'info, S, MAX_SELECTED_SHARDS, Unselected>;

/// Convenience alias for a `ShardSet` that already carries an active shard
/// selection.
pub type SelectedShards<'slice, 'info, S, const MAX_SELECTED_SHARDS: usize> =
    ShardSet<'slice, 'info, S, MAX_SELECTED_SHARDS, Selected>;

/// Allows constructing a `ShardSet` via `slice.into()` instead of calling
/// `ShardSet::new(slice)` explicitly.
impl<'slice, 'info, S, const MAX_SELECTED_SHARDS: usize> From<&'slice [AccountLoader<'info, S>]>
    for ShardSet<'slice, 'info, S, MAX_SELECTED_SHARDS, Unselected>
where
    S: ZeroCopy + Owner,
    'slice: 'info,
{
    #[inline]
    fn from(slice: &'slice [AccountLoader<'info, S>]) -> Self {
        ShardSet::from_loaders(slice)
    }
}
