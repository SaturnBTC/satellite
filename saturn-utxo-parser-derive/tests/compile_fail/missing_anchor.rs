use saturn_bitcoin_transactions::utxo_info::UtxoInfo;
use saturn_utxo_parser_derive::UtxoParser;
use anchor_lang::prelude::*;

// Dummy Accounts type deliberately missing the `missing` field referenced by the
// `anchor` attribute below. This should trigger a compile-time error.
#[derive(Debug, Accounts)]
struct DummyAccounts<'info> {
    acc: AccountInfo<'info>,
}

#[derive(UtxoParser)]
#[utxo_accounts(DummyAccounts)]
struct MissingAnchor {
    // The `anchor = missing` identifier does not exist on `DummyAccounts`; the macro
    // expansion will therefore reference a non-existent field and the compiler
    // must emit an error.
    #[utxo(anchor = missing)]
    utxo: UtxoInfo,
}

fn main() {}
