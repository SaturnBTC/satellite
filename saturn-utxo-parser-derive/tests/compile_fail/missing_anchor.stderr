error[E0609]: no field `missing` on type `&DummyAccounts<'info>`
  --> tests/compile_fail/missing_anchor.rs:18:21
   |
18 |     #[utxo(anchor = missing)]
   |                     ^^^^^^^ unknown field
   |
   = note: available field is: `acc`
